drivers/serial/uart_stm32.c

```diff
@@ -724,13 +724,13 @@ static inline void async_evt_rx_buf_release(struct uart_stm32_data *data)
 	async_user_callback(data, &evt);
 }
 
-static inline void async_timer_start(struct k_delayed_work *work,
+static inline void async_timer_start(struct k_work_delayable *work,
 				     int32_t timeout)
 {
 	if ((timeout != SYS_FOREVER_MS) && (timeout != 0)) {
 		/* start timer */
 		LOG_DBG("async timer started for %d ms", timeout);
-		k_delayed_work_submit(work, K_MSEC(timeout));
+		k_work_schedule(work, K_MSEC(timeout));
```

Comment:
Note that the recommended replacement for `k_delayed_work_submit()` is `k_work_reschedule()`.  There are cases where `k_work_schedule()` is appropriate, but be sure you understand the difference.

I'll continue with a review once it's confirmed this (and the other use) are correct for the way the driver works.

Reply:
Well, the difference between `k_work_reschedule` and `k_work_schedule` is not so clear for me. As we are in a starting function (the async_timer_start), I considered the work item could be simply scheduled there (not already scheduled). However the `k_work_reschedule` is present in other situations. Maybe reschedule could do _more_ than schedule.

Reply:
The difference is described [here](https://docs.zephyrproject.org/latest/reference/kernel/threads/workqueue.html#submitting-a-delayable-work-item).  If you know the work item is not pending at all points where it's scheduled, then the difference doesn't matter.  That there are unsynchronized and unchecked cancel operations means it's unclear that the work is not still pending or in  progress when  the start function is invoked.  It also seems plausible that `rx_cb` might schedule while the work is still pending.

The roughly equivalent prototype patch in #33924 used `reschedule` for all serial drivers.  If you are not sure, use `reschedule`.  If you are sure, then I'll finish the review.

Reply:
In the Best Practices section of #33104 Use the Right Scheduling Operation there is also a description of the difference which may be helpful.

Reply:
The `async_timer_start` is called during the isr, there, the k_work_reschedule shall be used.
It is also called during uart_stm32_async_tx. Considering that 'If the work item is not scheduled both APIs behave the same.' then I am changing to **k_work_reschedule** in all cases.